### ISSUE TITLE
Logstash reference mapping

### ISSUE THREAD

**Issue by dmuelle** [OpenLiberty/docs#1833]
Reference topic- [move info from this KC topic](https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/cwlp_analytics_collectors.html) , similar to [the Log and trace reference topic](https://www.openliberty.io/docs/ref/general/#logging.html).

SME: Don Bourne

**Comment by dmuelle** [comment]
Hi @donbourne, can you provide SME review of this reference doc?

https://draft-openlibertyio.mybluemix.net/docs/20.0.0.8/logstash-events-list.html

If possible, can you get back to me by Wednesday 7/29? Thanks!

**Comment by dmuelle** [comment]
@donbourne  does this table of HTTP access log event fields for Logstash collector need to be updated to contain the new custom fields that are available per LG-265 #1939 ? If so, does it make sense to add them in a separate table with a brief explanation that you can add them by specifying `jsonAccessLogFields="logFormat"`

**Comment by dmuelle** [comment]
new draft URL: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/logstash-events-list.html

**Comment by donbourne** [comment]
First 2 paragraphs are probably redundant given the descriptions in the Analyzing logs with elastic stack and forwarding logs with logstash topics

---

The Logstash collector feature captures log events at run time, breaks them into fields, and securely forwards them to the configured Logstash log collection server. You can configure the feature to send one or more of the following types of data:
->
The Logstash collector feature captures log events at run time, breaks them into fields, and securely forwards them to the configured Logstash log collection server. 

The following types of events are generated by the Logstash collector feature:

---

For consistency, should make list of events and their order match JSON log events reference list.  Only difference in the list is that logstash collector has GC events.

In general, all the event names and descriptions should match between the two topics.  

Table column names should match between the two topics. 

---

rather than have "common event fields" I suggest we follow JSON log events reference list and just show all fields in all events.

---

audit events is missing most of its fields (looking at the WL version it looks wrong too)

**Comment by donbourne** [comment]
in the nav bar,

Logstash event reference list -> Logstash collector events reference list

**Comment by dmuelle** [comment]
Hi @donbourne - just to confirm- all event names should match exactly between JSON log events and Logstash collector events? There are significant differences in the docs from KC.

For example, message events
`hostName` (logstash) vs `host`(JSON)
`threadId` vs `ibm_threadId`

many slight differences like this, esp `ibm_` before the field name in the JSON list but not on the Logstash list.
Just want to confirm that tables in these docs should be identical except for GC table in logstash list and that the JSON events list is the correct reference that the Logstash list should be patterned after

thanks

**Comment by donbourne** [comment]
I just meant the two topics should probably have consistency for look and feel.  The actual field names are different between json logging and logstashCollector, as you've seen. 

The list of event types shown at the top of both topics should be the same, give or take the GC events (I'm not saying the list of field names in those event types is the same between json logging and logstashCollector though).

**Comment by dmuelle** [comment]
Ok thanks for clarifying. For the missing audit event fields- where can I find those if not KC?

**Comment by Yushan-Lin** [comment]
hi @dmuelle , the missing audit event fields are the same as the JSON log events reference list. Anything starting with `ibm_audit_` applies to Logstash Collector as well.

**Comment by dmuelle** [comment]
Thanks @Yushan-Lin - when those event fields appear in Logstash Collector, do they retain the `ibm_audit_ ` prefix? I noticed that some of the other Logstash fields do not have an `ibm_` prefix, while their counterparts in the JSON reference list do

**Comment by Yushan-Lin** [comment]
Yup, they're the same.

**Comment by dmuelle** [comment]
@donbourne @Yushan-Lin this draft is updated and ready for review:

https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/logstash-events-list.html#_supported_audit_events_and_their_audit_data

Per Don's comments above, I've tried to make this match the [JSON events reference list](https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/json-log-events-list.html) as closely as possible. Other than the expected differences in the field names, there are a few discrepancies I noticed throughout that I wanted to double check:

## Event types

The callout list here is based on the KC. In this list, every type is listed as a "log event" except Garbage Collection and Audit
* Message log events
* Trace log events
* HTTP access log events
* First failure data capture (FFDC) log events
* Garbage collection events for IBM JDKs
* Audit Events

Are GC and Audit events typically not referred to as "log events" ? If they are I'll update the list so all the titles are the same format

## Message and trace log events
In both these tables, the JSON log events contains

`ext_thread | Thread name of the thread that is the source of the event`

While the Logstash list contains

`tags | The tags that are associated with the server from which the event originated`

I think these are essentially the same basic thing, but wanted to make sure that neither belongs in both tables

## FFDC Log events

The JSON reference list contains the following line, which was not in the Logstash table on KC- should it be included?

`message | The message from log record`

## HTTP Access log events
The JSON reference list contains the following lines, which were not in the Logstash table on KC- should they be included?

`ibm_requestHeader_{headername} | Header value from the request`
`ibm_responseHeader_{headername} | Header value from the response`
`ibm_responseHeader_{headername} | Header value from the response`
`ibm_requestStartTime | The start time of the request`
`ibm_remoteUserID | Remote user according to the WebSphere Application Server specific $WSRU header`
`ibm_accessLogDatetime | The time when the message to the access log is queued to be logged`
`ibm_remoteIP | Remote IP address, e.g. 127.0.0.1`
`ibm_bytesSent | Response size in bytes excluding headers`
`ibm_cookie_{cookiename} | Cookie value from the request`
`ibm_requestElapsedTime | The elapsed time of the request - millisecond accuracy, microsecond precision`

also, the following line is in the Logstash table, but not the JSON table, should it be?
`queryString | The string that represents the query string from the HTTP request, for example, color=blue&size=large`

## Audit events
Based on @Yushan-Lin's comment above, I directly copied these over from the JSON reference list. The onoly difference is that I added the Logstash "common" events to each table. Since these tables have already been reviewed for that topic, I left them otherwise as-is. Let me know if they need any edits.

Thanks! If possible, can you get back to me by Tues 8/18?

**Comment by Yushan-Lin** [comment]
@dmuelle for audit events it looks about the same except it's missing `threadId`

**Comment by dmuelle** [comment]
Thanks @Yushan-Lin, I've added `threadId` to all the audit tables.

**Comment by dmuelle** [comment]
I have a question re audit events and the Logstash collector feature.

In the introduction to this topic, we say
`The Logstash collector feature generates the following event types:`

`Audit events` is one of the listed types. Then, later, in the introduction to audit events (which I copied over from the JSON log events reference list), we say

`The Open Liberty Audit feature captures auditable events from the server runtime environment and applications.`

What is the relationship between the Logstash collector feature and the Audit feature? Does one of these two statements need to be edited? 

Thanks!

**Comment by Yushan-Lin** [comment]
@dmuelle The description is fine. Logstash Collector just sends Liberty events to the Logstash server. When audit feature is enabled, audit events are generated.

**Comment by donbourne** [comment]
@dmuelle , here are my answers to your questions...



## event types

The "log" part is somewhat redundant.  I think it would be cleaner to just drop it from all the events, as follows:

JSON:

- Message events
- Trace events
- First failure data capture (FFDC) events
- HTTP access events
- Audit events

logstash collector:

- Message events
- Trace events
- First failure data capture (FFDC) events
- HTTP access events
- Garbage collection events for IBM JDKs
- Audit events

## Message and trace log events

ext_thread and tags are not the same things.  ext_thread is the name of the thread that logged/traced.  tags come from the config and are used to decorate the event.

JSON logging does not have tags (there is no way to configure them)
ext_thread could appear in either logstashCollector or JSON logging for message and trace events

## FFDC log events

both json and logstash collector include "message" 

## HTTP Access log events

these can all appear in logstashCollector access log events.  the descriptions should be the same as for json, but the ibm_ prefix should not be used on logstashCollector side.

## Audit events

looks like you've already got an answer for this one.

**Comment by dmuelle** [comment]
Thanks Don. I've updated the draft to reflect your answers. Is this doc ready for sign-off or does it needs further edits? 

https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/logstash-events-list.html

**Comment by donbourne** [comment]
Hi @dmuelle , I have a few cleanup items to request...

in ffdc section (likely applies to this topic and the json reference topic):
    	
The message from the log record
->
The message from the exception that triggered the event


The class that emitted the FFDC entry
->
The class that emitted the FFDC event


The exception that was reported in the FFDC entry
->
The exception that was reported in the FFDC event

The thread ID of the FFDC incident
->
The thread ID of the FFDC event

The stack trace of the FFDC incident
->
The stack trace of the FFDC event

The incident details for the FFDC incident
->
The incident details for the FFDC event

---

please also move the following fields to top of each of the tables, and stick to the same order for these in each table:

- type
- datetime
- hostName
- wlpUserDir
- serverName
- sequence
- tags

**Comment by dmuelle** [comment]
Hi @donbourne - I've made these edits

- fixed ffdc descriptions as specified
- reordered common fields so they are the same in every table
- left a comment for @rw2513 in #210 to make the edits to the JSON ref list that you mentioned in this issue

Let me know if anything further is needed, thanks.

**Comment by donbourne** [comment]
spotted a couple more things...

- HTTP access event table 4 title should be - Table 4. HTTP access event fields
- HTTP access event table missing field requestFirstLine - First line of the request

**Comment by dmuelle** [comment]
Thanks Don, I've updated the HHTP access table title and added the missing field:

https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/logstash-events-list.html#_http_access_events

**Comment by donbourne** [comment]
Thanks @dmuelle , this looks good!

**Comment by dmuelle** [comment]
@lauracowen this doc is ready for your review

https://draft-openlibertyio.mybluemix.net/docs/20.0.0.10/logstash-events-list.html

**Comment by lauracowen** [comment]
This looks fine. You've worked with Don on it and I assume it complements the other events reference list topics. A couple of comments/style things below but I'll sign it off from my perspective. Thanks.

- The intro to the [Audit events section](https://draft-openlibertyio.mybluemix.net/docs/20.0.0.10/logstash-events-list.html#_supported_audit_events_and_their_audit_data) - might be worth checking what Richard's got in his now (he tweaked it slightly today) so all three topics are consistent in their wording/format here.
- Ha! You copy/pasted from Richard's topic - I spotted this typo in his this morning too: "JMX MBean Operations (JXM_MBEAN)"
- not sure if you want to address this here (it affects other topics too) but I'd rather we didn't have "Table 1." or "Figure 1." prefix on table/figure captions. It doesn't provide any value in online docs. It's a hangover from printed publishing practices where the writer would need to refer to a table/figure by number because it was printed on a different page. But in online docs it makes no sense to number them because the info is provided around the table/figure (and if you do need to refer up/down the page, I think you can just link up or down the page).

**Comment by dmuelle** [comment]
Thanks @lauracowen. I've updated the draft:

- The intro to the [Audit events section](https://draft-openlibertyio.mybluemix.net/docs/20.0.0.10/logstash-events-list.html#_supported_audit_events_and_their_audit_data) - might be worth checking what Richard's got in his now **updated**
- Ha! You copy/pasted from Richard's topic - I spotted this typo in his this morning too: "JMX MBean Operations (JXM_MBEAN)" **fixed :)**
- not sure if you want to address this here (it affects other topics too) but I'd rather we didn't have "Table 1." or "Figure 1." prefix on table/figure captions. **addressed at global level per our slack convo**

Builds are running super slow today but new draft should be posted at some point this afternoon

**Comment by dmuelle** [comment]
Requested peer review from @Rwalls1

**Comment by Rwalls1** [comment]
## Peer review
The topic looks good I just have a few comments:
### Server and host names in virtualized environments
- I think you could change “When Open Liberty servers are running” to “When Open Liberty servers run” to be more concise and use a more active voice
- I think the “hostName" and “serverName” terms should placed in code blocks because they reference code segments.
- I think the introductory sentences before each bulleted list should be reworded as the current construction seems a bit confusing, I think the sentences  are trying to say that the fields are set to whichever of the listed values appear first in the configuration but that doesn’t seem clear to me. Maybe change the sentences to something like, “The _____ field is set to the first available correlated value in the configuration. The configuration includes the following correlated values for the _____ field:"
- For the phrase, “set the CONTAINER_HOST and CONTAINER_NAME”, this constructions sounds like a command and should probably to be changed to something like: "you can set the CONTAINER_HOST and CONTAINER_NAME”
- The sentence that starts with “If you do not set these environment variables”, reads a bit awkwardly and also seems a bit wordy. Maybe change to something like, “If you do not set these environment variables, it is harder to determine the correlated container for events…
- For "use a command similar...”, the phrase sounds like a command and should be changed to something like, “you can use a command similar…
- For "the CONTAINER_HOST and CONTAINER_NAME are already set for you.”, I think you should add the appropriate nouns after "CONTAINER_HOST and CONTAINER_NAME” to maintain clarity and consistency throughout the topic.

**Comment by dmuelle** [comment]
Thanks for reviewing @Rwalls1. I made the following edits per your review:

- [x]  change “When Open Liberty servers are running” to “When Open Liberty servers run” 
- [x] the “hostName" and “serverName” terms should placed in code blocks 
- [x] the introductory sentences before each bulleted list should be reworded as the current construction seems a bit confusing, **I worked on this section to make it clearer and explain the value w/r/t the Logstash collector feature.**  
- [x] For the phrase, “set the CONTAINER_HOST and CONTAINER_NAME”, this constructions sounds like a command and should probably to be changed to something like: "you can set the CONTAINER_HOST and CONTAINER_NAME”
- [x] The sentence that starts with “If you do not set these environment variables”, reads a bit awkwardly and also seems a bit wordy. **rewrote this sentence**
- [x] For "use a command similar...”, the phrase sounds like a command and should be changed to something like, “you can use a command similar…
- [x] For "the CONTAINER_HOST and CONTAINER_NAME are already set for you.”, I think you should add the appropriate nouns after "CONTAINER_HOST and CONTAINER_NAME” to maintain clarity and consistency throughout the topic.

**Comment by dmuelle** [comment]
@chirp1  Note that this topic should not get final editorial review until after the JSON events reference list has been reviewed and approved and I make the changes to mirror any recommendations for content that is in both topics. That doc is tracked in #210

**Comment by dmuelle** [comment]
Draft URL: https://draft-openlibertyio.mybluemix.net/docs/latest/logstash-events-list.html

**Comment by dmuelle** [comment]
Edits for clarity and concreteness in #3102 

- reduce wordiness
- use technical terms consistently
- Make elements parallel
- USE CONSISTENT MARKUP TAGGING

**Comment by dmuelle** [comment]
Doc is now on vNext branch, will publish with 20.0.0.12 on 11.20

**Comment by dmuelle** [comment]
https://openliberty.io/docs/latest/logstash-events-list.html

**Comment by dmuelle** [comment]
re-reviewed for editing checklist concerns- no futher edits needed at this time.
